File: src/main/java/com/netsensia/rivalchess/engine/core/eval/StaticExchangeEvaluatorPremium.java
package com.netsensia.rivalchess.engine.core.eval;

import com.netsensia.rivalchess.config.Limit;
import com.netsensia.rivalchess.model.Colour;
import com.netsensia.rivalchess.engine.core.EngineChessBoard;
import com.netsensia.rivalchess.engine.core.type.EngineMove;
import com.netsensia.rivalchess.exception.InvalidMoveException;

import java.util.ArrayList;
import java.util.List;

public class StaticExchangeEvaluatorPremium implements StaticExchangeEvaluator {

    public int staticExchangeEvaluation(EngineChessBoard board, EngineMove move) throws InvalidMoveException {
        final int captureSquare = move.compact & 63;
        final int materialBalance = materialBalanceFromMoverPerspective(board);

        if (board.makeMove(move)) {
            final int seeValue = -seeSearch(board, captureSquare) - materialBalance;
            board.unMakeMove();
            return seeValue;
        }

        return -Integer.MAX_VALUE;
    }

    public int seeSearch(EngineChessBoard board, int captureSquare) throws InvalidMoveException {

        final int materialBalance = materialBalanceFromMoverPerspective(board);

        int bestScore = materialBalance;

        List<EngineMove> moves = getCaptureMovesOnSquare(board, captureSquare);

        for (EngineMove move : moves) {

            if (board.makeMove(move)) {
                final int seeScore = -seeSearch(board, captureSquare);
                board.unMakeMove();
                if (seeScore > bestScore) {
                    bestScore = seeScore;
                }
            }
        }
        
        return bestScore;
    }

    public int materialBalanceFromMoverPerspective(EngineChessBoard board) {
        final int whiteMaterial = board.getWhitePawnValues() + board.getWhitePieceValues();
        final int blackMaterial = board.getBlackPawnValues() + board.getBlackPieceValues();

        if (board.getMover() == Colour.WHITE) {
            return whiteMaterial - blackMaterial;
        }

        return blackMaterial - whiteMaterial;
    }

    public List<EngineMove> getCaptureMovesOnSquare(EngineChessBoard board, int captureSquare) {
        int[] moves = new int[Limit.MAX_LEGAL_MOVES.getValue()];
        final boolean includeChecks = false;
        board.setLegalQuiesceMoves(moves, includeChecks);
        List<EngineMove> moveList = new ArrayList<>();

        int moveNum = 0;
        int move = moves[moveNum];
        while (move != 0) {
            if ((move & 63) == captureSquare) {
                moveList.add(new EngineMove(move));
            }
            move = moves[++moveNum];
        }
        return moveList;
    }
}

File: src/main/java/com/netsensia/rivalchess/engine/core/eval/StaticExchangeEvaluatorPremium.kt
package com.netsensia.rivalchess.engine.core.eval

import com.netsensia.rivalchess.config.Limit
import com.netsensia.rivalchess.engine.core.board.EngineBoard
import com.netsensia.rivalchess.engine.core.type.EngineMove
import com.netsensia.rivalchess.exception.InvalidMoveException
import com.netsensia.rivalchess.model.Colour
import java.util.*

class StaticExchangeEvaluatorPremium : StaticExchangeEvaluator {
    @Throws(InvalidMoveException::class)
    override fun staticExchangeEvaluation(board: EngineBoard, move: EngineMove): Int {
        val captureSquare = move.compact and 63
        val materialBalance = materialBalanceFromMoverPerspective(board)
        if (board.makeMove(move)) {
            val seeValue = -seeSearch(board, captureSquare) - materialBalance
            board.unMakeMove()
            return seeValue
        }
        return -Int.MAX_VALUE
    }

    @Throws(InvalidMoveException::class)
    fun seeSearch(board: EngineBoard, captureSquare: Int): Int {

        val moves = getCaptureMovesOnSquare(board, captureSquare)

        var bestScore = materialBalanceFromMoverPerspective(board)

        for (move in moves) {
            if (board.makeMove(move)) {
                val seeScore = -seeSearch(board, captureSquare)
                board.unMakeMove()
                bestScore = seeScore.coerceAtLeast(bestScore)
            }
        }
        return bestScore
    }

    private fun materialBalanceFromMoverPerspective(board: EngineBoard): Int {
        val whiteMaterial = board.whitePawnValues + board.whitePieceValues
        val blackMaterial = board.blackPawnValues + board.blackPieceValues
        return if (board.mover == Colour.WHITE) {
            whiteMaterial - blackMaterial
        } else blackMaterial - whiteMaterial
    }

    private fun getCaptureMovesOnSquare(board: EngineBoard, captureSquare: Int): List<EngineMove> {
        val moves = board.getQuiesceMoveArray(false)
        val moveList: MutableList<EngineMove> = ArrayList()
        var moveNum = 0
        var move = moves[moveNum]
        while (move != 0) {
            if (move and 63 == captureSquare) {
                moveList.add(EngineMove(move))
            }
            move = moves[++moveNum]
        }
        return moveList
    }
}